src/csharp/Grpc.Core.Api/ServerCallContext.cs

```diff
@@ -113,6 +116,23 @@ public WriteOptions WriteOptions
         /// </summary>
         public AuthContext AuthContext => AuthContextCore;
 
+        /// <summary>
+        /// Gets a dictionary that can be used by the various interceptors and handlers of this
+        /// call to store arbitrary state.
+        /// </summary>
+        public virtual IDictionary<object, object> UserState
```

Comment:
what are the arguments for using  IDictionary<object, object>  over Get<T>() and Set<T>()? I'm not a big fan of allowing the key to be of arbitrary type (and perhaps users using one that's not well suited to be a key in a dictionary). Also, I think not having a generic version of Get<T>() make retrieving typed values a bit clumsy because the user needs to use an explicit cast to get hold of the value.

Btw, do we expect the pattern `UserState.Add(typeof(MyData), instanceOfMyData)` to be common?

Reply:
I'm happy to implement `Get<T>`/`Set(T)`. As the maintainer of the library and the overall API shape, I'll defer to your for a final choice here.

Do you want me to implement `Get<T>`/`Set(T)` so we can compare the APIs?

Pros of `IDictionary<,>`:

* Can store multiple values of the same type without needing to introduce wrapper types.
* Works with any existing dictionary-based methods.
* Can differentiate between not set and a stored null/default. I'd expect `Get<T>()` to return null/default if there hadn't been a call to `Set(T)` yet.
* Very minor: `Dictionary<,>` is already implemented and tested. :-)

Cons:

* Explicit cast needed.
* Can differentiate between not set and a stored null/default. (Yes, both a pro and a con.)

My guess is that many people would use string constants for keys. For those who want to be robust against inadvertent collisions with arbitrary libraries, they'd probably use a private object instance (like is done for locking).

Reply:
Exposing an `IDictionary<object, object>` has been fine for ASP.NET Core.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.items